simple_transformers/modality_processors.py

```diff
@@ -0,0 +1,301 @@
+from abc import ABC, abstractmethod
+import math
+import numpy as np
+from PIL.Image import Image, fromarray
+import torch as th
+import torch.nn as nn
+from transformers import CLIPTokenizer, CLIPImageProcessor
+
+from types import SimpleNamespace
+from typing import Union, List, Tuple, Any
+
+"""
+All processors can be used for encoding. 
+Only processors with 'output_embeddings_to_logits' and 'logits_to_output' functions can be used for decoding.
+Currently this only includes TextProcessor and ActionProcessor
+"""
+
+class Processor(nn.Module, ABC):
+    def __init__(self, config: SimpleNamespace, **kwargs):
+        """
+        :param config: Config namespace to hold model parameters.
+                       See simple_transformers.config.yaml and simple_transformers.utils.py
+        :param kwargs: kwargs required from dataset (e.g. max trajectory length)
+        """
+        super().__init__()
+        self.config = config
+        self.LayerNorm = nn.LayerNorm(config.d_model, eps=config.layer_norm_eps)
+        self.dropout = nn.Dropout(config.dropout_prob)
+
+    @abstractmethod
+    def forward(self, modality: Union[str, np.array, Image, Any], att_mask: Union[np.array, None]) -> \
+                Tuple[th.Tensor, th.Tensor]:
+        """
+        Takes in the modality you want to process. Type and format of input will vary based on modality. For example,
+        a text processor could take in a string whereas an image processor could take in a PIL.Image.
+        Must return input embeddings and an attention mask to feed into a transformer.
+        NOTE: Attention mask follows pytorch's format, not Hugging faces format. i.e. 0->unmasked, 1->masked
+        modality: Input modality
+        att_mask: Attention mask on modality. Some modality processors don't require this as they will generate it
+                  themselves. In this case, pass in None
+        Returns: Fully formed input tokens for a transformer (including all position/modality embeddings)
+                 Attention mask for the input tokens. Both should be th.tensors on the correct device
+        """
+        pass
+
+
+class TextProcessor(Processor):
+    """
+    Processes string text into input embeddings. Uses a pretrained tokenizer from huggingface.
+    Attention mask is created by tokenizer..
+    REQUIRES: 'max_text_length' kwarg
```

Comment:
I think it would be useful to have this check enforced in the code. You could provide a checker as part of the `super().__init__()` something like:
```python
class Processor(nn.Module, ABC):
    def __init__(self, config: SimpleNamespace, **kwargs):
        """
        :param config: Config namespace to hold model parameters.
                       See simple_transformers.config.yaml and simple_transformers.utils.py
        :param kwargs: kwargs required from dataset (e.g. max trajectory length)
        """
        super().__init__()
        check_required_kwargs(kwargs)
        ...
        

    @abstractmethod
    def forward(self, modality: Union[str, np.array, Image, Any], att_mask: Union[np.array, None]) -> \
                Tuple[th.Tensor, th.Tensor]:
        ...
        pass

    @property
    @abstractmethod
    def required_attributes(self):
        pass

    def check_required_kwargs(self)...
        # check against self.required_attributes

class TextProcessor(Processor):
    ...
    def required_attributes(self):
        return ['max_text_length']
```